tests/e2e/ctl_v3_member_test.go

```diff
@@ -309,3 +312,104 @@ func TestRemoveNonExistingMember(t *testing.T) {
 	err = epc.Restart(ctx)
 	assert.NoError(t, err)
 }
+
+func TestMemberReplace(t *testing.T) {
+    e2e.BeforeTest(t)
+    ctx := context.Background()
+
+    epc, err := e2e.NewEtcdProcessCluster(ctx, t)
+    assert.NoError(t, err)
+    defer epc.Close()
+
+	memberId := rand.Int() % len(epc.Procs)
+	member := epc.Procs[memberId]
+	var endpoints []string
+	for i := 1; i < len(epc.Procs); i++ {
+		endpoints = append(endpoints, epc.Procs[(memberId+i)%len(epc.Procs)].EndpointsGRPC()...)
+	}
+	cc, err := e2e.NewEtcdctl(epc.Cfg.Client, endpoints)
+	assert.NoError(t, err)
+
+	c := epc.Etcdctl()
+	memberID, found, err := getID(ctx, c, member.Config().Name)
+	assert.NoError(t, err)
+	assert.Equal(t, found, true, "Member not found")
+
+	// Need to wait health interval for cluster to accept member changes
+	time.Sleep(etcdserver.HealthInterval)
+	_, err = c.MemberRemove(ctx, memberID)
+	assert.NoError(t, err)
+	_, found, err = getID(ctx, c, member.Config().Name)
+	assert.NoError(t, err)
+	assert.Equal(t, found, false, "Expected member to be removed")
+
+	t.Logf("Removing member %s", member.Config().Name)
+	for member.IsRunning() {
+		member.Kill()
+		err = member.Wait(ctx)
+		if err != nil && !strings.Contains(err.Error(), "unexpected exit code") {
+			t.Error("failed to kill the process. Err: %w", err)
+		}
+	}
+	t.Logf("Removing member %s data", member.Config().Name)
+	err = os.RemoveAll(member.Config().DataDirPath)
+	assert.NoError(t, err)
+	
+	t.Logf("Adding member %s back", member.Config().Name)
+	removedMemberPeerUrl := member.Config().PeerURL.String()
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fail()
+		default:
+		}
+		reqCtx, cancel := context.WithTimeout(ctx, time.Second)
+		_, err = cc.MemberAdd(reqCtx, member.Config().Name, []string{removedMemberPeerUrl})
```

Comment:
Very interesting. I have used this test to validate https://github.com/etcd-io/etcd/pull/17056, looks like the issue is triggered by the immidiate `MemberAdd` request. When I removed the for loop and added a `time.Sleep(time.Second)` issue no longer reproduces.

This makes me think we have risk of potential race condition on membership changes that was is not covered by test at all. cc @ahrtr

Reply:
Initially, I was using `assert.NoError()`, which allowed the test to continue executing even when errors occurred in the `MemberAdd` request. Upon logging the error message, I found that

>     ctl_v3_member_no_proxy_test.go:92: Adding member TestMemberReplace-test-2 back failed: match not found.  Set EXPECT_DEBUG for more info Errs: [unexpected exit code [1] after running [/home/zhoujian/etcd-test/etcd/bin/etcdctl --endpoints=http://localhost:20000,http://localhost:20005 member add TestMemberReplace-test-2 --peer-urls http://localhost:20011 -w json]], last lines:
>         {"level":"warn","ts":"2023-12-12T14:35:14.017056+0800","logger":"etcd-client","caller":"v3@v3.6.0-alpha.0/retry_interceptor.go:65","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001c2780/localhost:20000","method":"/etcdserverpb.Cluster/MemberAdd","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: Peer URLs already exists"}
>         Error: etcdserver: Peer URLs already exists

In your case, a single `MemberAdd` request might fail, yet the test continues. Consequently, the cluster doesn't re-add the member, and it loops to locate the corresponding member ID until a timeout occurs. This divergence could arise from issue #17017, which might not get the updated `membersMap` timely to check peer urls.